src/style-installer.js

```diff
@@ -14,21 +14,70 @@
  * limitations under the License.
  */
 
-import {dev, rethrowAsync} from './log';
 import {Services} from './services';
+import {dev, rethrowAsync} from './log';
+import {map} from './utils/object';
 import {setStyles} from './style';
+import {waitForBody} from './dom';
 import {waitForServices} from './render-delaying-services';
 
-
+const TRANSFORMER_PROP = '__AMP_CSS_TR';
+const STYLE_MAP_PROP = '__AMP_CSS_SM';
 const bodyVisibleSentinel = '__AMP_BODY_VISIBLE';
 
+
 /**
- * Adds the given css text to the given document.
+ * Adds the given css text to the given ampdoc.
  *
  * The style tags will be at the beginning of the head before all author
  * styles. One element can be the main runtime CSS. This is guaranteed
  * to always be the first stylesheet in the doc.
  *
+ * @param {!./service/ampdoc-impl.AmpDoc} ampdoc The ampdoc that should get the new styles.
+ * @param {string} cssText
+ * @param {function(!Element)} cb Called when the new styles are available.
+ *     Not using a promise, because this is synchronous when possible.
+ *     for better performance.
+ * @param {boolean=} opt_isRuntimeCss If true, this style tag will be inserted
+ *     as the first element in head and all style elements will be positioned
+ *     after.
+ * @param {string=} opt_ext
+ * @return {!Element}
+ */
+export function installStylesForDoc(
+    ampdoc, cssText, cb, opt_isRuntimeCss, opt_ext) {
+  const rootNode = ampdoc.getRootNode();
+  const cssRoot = ampdoc.getHeadNode();
+  const style = insertStyleElement(
+      cssRoot,
+      maybeTransform(cssRoot, cssText),
+      opt_isRuntimeCss || false,
+      opt_ext || null);
+
+  // Styles aren't always available synchronously. E.g. if there is a
+  // pending style download, it will have to finish before the new
+  // style is visible.
+  // For this reason we poll until the style becomes available.
+  // Sync case.
+  if (styleLoaded(rootNode, style)) {
+    cb(style);
```

Comment:
Is this a significant perf improvement compared to microtask? Mixing sync and async...